doc/man3/SSL_shutdown.pod

```diff
@@ -21,161 +21,242 @@ SSL_shutdown, SSL_shutdown_ex - shut down a TLS/SSL or QUIC connection
+
+=head2 Locally-Initiated Shutdown
+
+Calling SSL_shutdown() on a SSL/TLS or DTLS SSL object initiates the shutdown
+process and causes OpenSSL to try to send a close_notify shutdown alert to the
+peer. The shutdown process will then be considered completed once the peer
+responds in turn with a close_notify shutdown alert message.
+
+Calling SSL_shutdown() only closes the write direction of the connection; the
+read direction is closed by the peer. Once SSL_shutdown() is called,
+L<SSL_write(3)> can no longer be used, but L<SSL_read(3)> may still be used
+until the peer decides to close the connection in turn. The peer might
+continuing sending data for some period of time before honouring the local
+application's shutdown request.
+
+=head2 Remotely-Initiated Shutdown
+
+If the peer was the first to initiate the shutdown process by sending a
+close_notify alert message, an application will be notified of this as an EOF
+condition when calling
+L<SSL_read(3)> (i.e., L<SSL_read(3)> will fail and L<SSL_get_error(3)> will
+return B<SSL_ERROR_ZERO_RETURN>), after any application data sent by the peer
+prior to initiating the shutdown has been read. An application should handle
+this condition by calling SSL_shutdown() to respond with a close_notify alert in
+turn, completing the shutdown process, though it may choose to write additional
+application data using L<SSL_write(3)> before doing so. If an application does
+not call SSL_shutdown() in this case, a close_notify alert will not be sent and
+the behaviour will not be fully standards compliant.
+
+=head2 Shutdown Lifecycle
+
+Regardless of whether a shutdown was initiated locally or by the peer, if the
+underlying BIO is blocking, a call to SSL_shutdown() will return either once
+close_notify alert messages have both been sent to and received from the peer,
+and the shutdown process is thereby completed (returning 1), or if an error
+occurs.
+
+If the underlying BIO is nonblocking and the shutdown process is not yet
+complete (for example, because a close_notify alert message has not yet been
+received from the peer, or because a close_notify alert message needs to be sent
+but would currently block), SSL_shutdown() returns 0 to indicate that the
+shutdown process is still ongoing; in this case, a call to L<SSL_get_error(3)>
+will yield B<SSL_ERROR_WANT_READ> or B<SSL_ERROR_WANT_WRITE>.
+
+An application can then detect completion of the shutdown process by calling
+SSL_shutdown() again repeatedly until it returns 1, indicating that the shutdown
+process is complete (with a close_notify alert having both been sent and
+received). However, the preferred method of waiting for the shutdown to complete
+is to use L<SSL_read(3)> until L<SSL_get_error(3)> indicates EOF by returning
+B<SSL_ERROR_ZERO_RETURN>. This ensures any data received immediately before the
+peer's close_notify alert is still provided to the application.
```

Comment:
As far as I know, the reason for that recommendation is to be able to get session tickets in TLS 1.3, which are not part of the handshake and you wouldn't get them if you never called SSL_read(). (Changing that behaviour, so that you still get them on SSL_shutdown() might be a good idea.)

Reply:
I guess SSL_shutdown() also does it based on other text, so I'm not sure any more why we recommend it.

Reply:
In #5823 I added:
`If the peer did send data it needs to be processed by calling SSL_read()
before calling SSL_shutdown() a second time.
SSL_read() will indicate the end of the peer data by returning <= 0
and SSL_get_error() returning SSL_ERROR_ZERO_RETURN.
It is recommended to call SSL_read() between SSL_shutdown() calls.
`

The reason for adding that is that SSL_shutdown() returns an error when not all data was read.

In #7188 I've reworded things again, but probably didn't actually make it more clear.

Reply:
>The reason for adding that is that SSL_shutdown() returns an error when not all data was read.

Hmm, how does this work? The two directions of flow can be closed in either order.

So if a local application writes data using SSL_write, then calls SSL_shutdown, the peer is entitled to reply with other data before shutting down in turn. Then the local application calls SSL_read until SSL_ERROR_ZERO_RETURN, and can confirm the shutdown if it wishes by calling SSL_shutdown() again, which then returns 1.

But what if a local application writes data using SSL_write, reads all pending data using SSL_read, then (race condition) the peer sends some more data immediately before a call to SSL_shutdown()?

Does SSL_shutdown() error if there is residual unread data when initiating a shutdown (would return 0 but for the unread data error) or when completing a shutdown (would return 1 but for the unread data error)?